### PR TITLE
[2.8] Add "How to Use Multiple Guard Authenticators" cookbook documentation

### PR DESCRIPTION
**Pull request by marek-pietrzak-tg**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes (https://github.com/symfony/symfony/pull/14673) |
| Applies to | `2.8` onwards |

Hi guys,
this is my first contribution to the symfony docs.
During my preparations for the Symfony Guard component workshops I have spent some time trying to figure out the problem described here.
I hope this cookbook entry will help others save their time.

cc @weaverryan 

Thanks!

### CODE REVIEW

File: cookbook/security/multiple_guard_authenticators.rst [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+========================================
+
+Guard authentication component allows you to easily use many different authenticators at a time.
+
+An entry point is a service id (of one of your authenticators) whose start()
+method should be called when an anonymous user hits a page that requires authentication.
+
+Multiple authenticators with shared entry point
+-----------------------------------------------
+Let's have an example of two authenticators: one based on login form, another one on facebook login.
```
Comment (dosten) [inline cookbook/security/multiple_guard_authenticators.rst]: Missing empty line before this line

**Comment by mheki** [comment]
Any thoughts on this?

File: cookbook/security/multiple_guard_authenticators.rst [hunk 2]
```diff
@@ -0,0 +1,168 @@
+How to Use Multiple Guard Authenticators
+========================================
+
+Guard authentication component allows you to easily use many different authenticators at a time.
```
Comment (xabbuh) [inline cookbook/security/multiple_guard_authenticators.rst]: The Guard [...]

File: cookbook/security/multiple_guard_authenticators.rst [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+An entry point is a service id (of one of your authenticators) whose start()
```
Comment (xabbuh) [inline cookbook/security/multiple_guard_authenticators.rst]: please enclose `start()` with literals (double backticks)

File: cookbook/security/multiple_guard_authenticators.rst [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+method should be called when an anonymous user hits a page that requires authentication.
+
+Multiple authenticators with shared entry point
```
Comment (xabbuh) [inline cookbook/security/multiple_guard_authenticators.rst]: due to how we capitalise headlines this should be "Multiple Authenticators with a Shared Entry Point"

File: cookbook/security/multiple_guard_authenticators.rst [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+-----------------------------------------------
+
+Let's have an example of two authenticators: one based on login form, another one on facebook login.
+Both authenticators entry points redirect user to the same login page.
+However, in your configuration you have to explicitly say which entry point you want to use.
```
Comment (xabbuh) [inline cookbook/security/multiple_guard_authenticators.rst]: We avoid the first person perspective in the docs. Thus, I would reword this paragraph to something like the following:

> Sometimes you want to offer your users different authentication mechanisms like a form login and a Facebook login while both entry points redirect the user to the same login page. However, [...]

File: cookbook/security/multiple_guard_authenticators.rst [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                            'app.facebook_connect_authenticator'
+                        ),
+                    ),
+                ),
+            ),
+        ));
+
+There is one limitation with this approach - you have to use exactly one entry point.
+
+Multiple authenticators with separate entry points
```
Comment (xabbuh) [inline cookbook/security/multiple_guard_authenticators.rst]: Multiple Authenticators with separate Entry Points

File: cookbook/security/multiple_guard_authenticators.rst [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+--------------------------------------------------
+
+Let's now have an example of two different authenticators: one based on login form, another one on an API token.
+When user hits secured area he should be redirected to the login page.
+Also when user hits an API endpoint, he should get a relevant API response.
+
+Solution for this use case is to provide guard authenticators in two separate firewalls.
+
+This is an example of your configuration:
```
Comment (xabbuh) [inline cookbook/security/multiple_guard_authenticators.rst]: I would reword this to something like this:

> However, there are use cases where you have authenticators that protect different parts of your application. For example, you have a login form that protects the secured area of your application front-end and API end points that are protected with API tokens. As you can only configure one entry point  per firewall the solution is to split the configuration into two separate firewalls:

**Comment by xabbuh** [comment]
I am really sorry that we missed to review your pull request for such a long time @mheki! This looks really great and I just left you some minor comments.

File: cookbook/security/multiple_guard_authenticators.rst [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xmlns:srv="http://symfony.com/schema/dic/services"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <config>
+                <!-- ... -->
+                <firewall name="default">
+                    <anonymous />
+                    <guard entry_point="app.form_login_authenticator">
```
Comment (WouterJ) [inline cookbook/security/multiple_guard_authenticators.rst]: should be `entry-point` (XML conventions are to not use underscores and the XmlFileLoader takes care of transforming all dashes into underscores)

File: cookbook/security/multiple_guard_authenticators.rst [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+                            - app.api_token_authenticator
+                default:
+                    anonymous: ~
+                    guard:
+                        authenticators:
+                            - app.form_login_authenticator
+            access_control:
+                - { path: ^/login, roles: IS_AUTHENTICATED_ANONYMOUSLY }
+                - { path: ^/api, roles: ROLE_API_USER }
+                - { path: ^/, roles: ROLE_ADMIN }
```
Comment (WouterJ) [inline cookbook/security/multiple_guard_authenticators.rst]: this means the whole site can only be accessed by its admins. What about changing it to `ROLE_USER` or `^/admin`?

File: cookbook/security/multiple_guard_authenticators.rst [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        ));
+
+There is one limitation with this approach - you have to use exactly one entry point.
+
+Multiple authenticators with separate entry points
+--------------------------------------------------
+
+Let's now have an example of two different authenticators: one based on login form, another one on an API token.
+When user hits secured area he should be redirected to the login page.
+Also when user hits an API endpoint, he should get a relevant API response.
```
Comment (WouterJ) [inline cookbook/security/multiple_guard_authenticators.rst]: I would replace "Also when user hits[...]" with "But when a user hits [...]". We also use the gender-neutral form in the docs, please replace "he should get a relevant API response" with "they should get [...]".

File: cookbook/security/multiple_guard_authenticators.rst [hunk 11]
```diff
@@ -0,0 +1,168 @@
+How to Use Multiple Guard Authenticators
+========================================
+
+Guard authentication component allows you to easily use many different authenticators at a time.
+
+An entry point is a service id (of one of your authenticators) whose start()
+method should be called when an anonymous user hits a page that requires authentication.
```
Comment (WouterJ) [inline cookbook/security/multiple_guard_authenticators.rst]: While I know that this is the official description of the entry point, I think we have to change it a bit. An anonymous user can be confused with the `AnonymousToken`. This anonymous token is created by the authentication process (a user is authenticated as being Anonymous).

Afaik, entry points are called when there is no token at all. In such case, the task of an entry point is to start the authentication procedure (e.g. redirect to the login page or return some header that triggers digest authentication).

**Comment by mheki** [comment]
Thanks for the review @xabbuh @WouterJ I've changed the docs according to your suggestions. 
Could you take a look again please?

**Comment by WouterJ** [comment]
Looks great, thanks for submitting the docs and your quick response @mheki! :+1: 

I think we should add a ``.. versionadded::`` directive indicating Guard was added in 2.8 directly after the headline, but we can add this after the merge.

**Comment by mheki** [comment]
Hmmm... I've just seen your commit `Removed all 2.x versionadded directives` made couple days ago...

**Comment by WouterJ** [comment]
@mheki yeah, for the 3.0 + master branch. I think it still makes sense for the 2.8 branch.

**Comment by mheki** [comment]
:+1:  added `versionadded`

**Comment by WouterJ** [comment]
Nice, thanks! :)

**Comment by xabbuh** [comment]
Thank you @mheki for this nice pull request and sorry again for  the delay when reviewing it. I have squashed your commits before merging them into the `2.8` branch. That's why the PR is displayed as closed instead of merged (but you will still get the full credit for your contribution).
